### ISSUE TITLE
Compile errors in giac with c++ 17 on Windows & MacOS

### ISSUE THREAD

**Issue by tobiasdiez** [sagemath/sage#38985]
### Steps To Reproduce

Remove the line `# distutils: extra_compile_args = -std=c++11` in `src\sage\libs\giac\giac.pyx`. Then compile with MSVC.

### Expected Behavior

Compiles fine

### Actual Behavior

```
  In file included from /Users/runner/miniconda3/envs/sage/include/giac/giac.h:5:
  In file included from /Users/runner/miniconda3/envs/sage/include/giac/poly.h:26:
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:341:10: error: no template named 'pointer_to_binary_function' in namespace 'std'
      std::pointer_to_binary_function < const monomial<T> &, const monomial<T> &, bool> strictly_greater ;
      ~~~~~^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:342:28: error: no template named 'pointer_to_binary_function' in namespace 'std'
      sort_helper(const std::pointer_to_binary_function < const monomial<T> &, const monomial<T> &, bool> is_strictly_greater):strictly_greater(is_strictly_greater) {};
                        ~~~~~^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:343:41: error: no member named 'ptr_fun' in namespace 'std'
      sort_helper():strictly_greater(std::ptr_fun<const monomial<T> &, const monomial<T> &, bool>(m_lex_is_strictly_greater<T>)) {};
                                     ~~~~~^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:343:49: error: expected expression
      sort_helper():strictly_greater(std::ptr_fun<const monomial<T> &, const monomial<T> &, bool>(m_lex_is_strictly_greater<T>)) {};
                                                  ^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:343:70: error: expected expression
      sort_helper():strictly_greater(std::ptr_fun<const monomial<T> &, const monomial<T> &, bool>(m_lex_is_strictly_greater<T>)) {};
                                                                       ^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:343:95: error: expected '(' for function-style cast or type construction
      sort_helper():strictly_greater(std::ptr_fun<const monomial<T> &, const monomial<T> &, bool>(m_lex_is_strictly_greater<T>)) {};
                                                                                            ~~~~^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:680:18: error: no template named 'pointer_to_binary_function' in namespace 'std'
               const std::pointer_to_binary_function < const monomial<T> &, const monomial<T> &, bool> m_is_strictly_greater
                     ~~~~~^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:729:44: error: no template named 'pointer_to_binary_function' in namespace 'std'
      typedef std::map< index_t,T,const std::pointer_to_binary_function < const index_m &, const index_m &, bool> > application;
                                        ~~~~~^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:730:30: error: no type named 'ptr_fun' in namespace 'std'
      application produit(std::ptr_fun(is_strictly_greater));
                          ~~~~~^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:737:14: error: 'application' (aka 'int') is not a class, namespace, or enumeration
      typename application::iterator prod_it,prod_itend;
               ^
  /Users/runner/miniconda3/envs/sage/include/giac/monomial.h:754:29: error: use of undeclared identifier 'prod_itend'; did you mean 'prod_it'?
      prod_it=produit.begin(),prod_itend=produit.end();
```

### Environment

- **OS**: 
- **Sage Version**: 

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

**Comment by user202729** [comment]
`giac/first.h`:

```cpp
#if __cplusplus >= 201103L
// post-c++11 functional headers changes, thanks to George Huebner
#define CPP11
#endif
```

`giac/monomial.h`:

```cpp
#ifdef CPP11
    std::function<bool(const monomial<T> &, const monomial<T> &)> strictly_greater ;
    sort_helper(const std::function<bool(const monomial<T> &, const monomial<T> &)> is_strictly_greater):strictly_greater(is_strictly_greater) {};
    sort_helper():strictly_greater(m_lex_is_strictly_greater<T>) {};
#else
    std::pointer_to_binary_function < const monomial<T> &, const monomial<T> &, bool> strictly_greater ;
    sort_helper(const std::pointer_to_binary_function < const monomial<T> &, const monomial<T> &, bool> is_strictly_greater):strictly_greater(is_strictly_greater) {};
    sort_helper():strictly_greater(std::ptr_fun<const monomial<T> &, const monomial<T> &, bool>(m_lex_is_strictly_greater<T>)) {};
#endif
```

In principle, the error cannot happen if everything are configured correctly.

I cannot reproduce the error on my machine (either replacing `11` with `17`, or delete the line entirely as you suggested).

(I'm compiling with meson however, which if I recalled correctly might not parse the `# distutils` file headers, but by enabling `--verbose` I see it is compiling with C++17 by default.)

**Comment by tobiasdiez** [comment]
Thanks @user202729 for looking into this, and sorry for not providing not enough information in the first place.

You are right, I'm using MSVC and thus the check in `giac/first.h` is not sufficient.

@parisseb could you please change the line
```
#if __cplusplus >= 201103L
```
to 
```
#if __cplusplus >= 201103L || (defined(_MSVC_LANG) && _MSVC_LANG >= 201103L)
```

Thanks!

**Comment by tobiasdiez** [comment]
Same error occurs also on MacOS: https://github.com/sagemath/sage/actions/runs/12579685480/job/35060278092?pr=38879#step:8:4735

**Comment by user202729** [comment]
Might be conda issue this time.

On my machine

```
$ cat $CONDA_PREFIX/conda-meta/giac-1.9.0.21-h673759e_1.json 

→

...
  "extracted_package_dir": "/home/user202729/miniforge3/pkgs/giac-1.9.0.21-h673759e_1",
  "files": [
...
    "include/giac/first.h",
...

$ cat /home/user202729/miniforge3/pkgs/giac-1.9.0.21-h673759e_1/include/giac/first.h

...
[no #define CPP11 seen]
```

tl;dr version of giac on conda-forge is too old that it doesn't include the `#define CPP11` patch.

https://github.com/conda-forge/giac-feedstock
